core/field.js

```diff
@@ -64,1155 +64,1194 @@ goog.require('Blockly.Gesture');
+    }
   }
-  this.setValue(state);
-};
 
-/**
- * Returns a stringified version of the XML state, if it should be used.
- * Otherwise this returns null, to signal the field should use its own
- * serialization.
- * @param {*} callingClass The class calling this method.
- *     Used to see if `this` has overridden any relevant hooks.
- * @return {?string} The stringified version of the XML state, or null.
- * @protected
- */
-Field.prototype.saveLegacyState = function(callingClass) {
-  if (callingClass.prototype.saveState === this.saveState &&
-      callingClass.prototype.toXml !== this.toXml) {
-    const elem = utilsXml.createElement('field');
-    elem.setAttribute('name', this.name || '');
-    const text = Xml.domToText(this.toXml(elem));
-    return text.replace(
-        ' xmlns="https://developers.google.com/blockly/xml"', '');
+  /**
+   * Set whether this field's value can be changed using the editor when the
+   *     source block is editable.
+   * @param {boolean} enabled True if enabled.
+   */
+  setEnabled(enabled) {
+    this.enabled_ = enabled;
+    this.updateEditable();
   }
-  // Either they called this on purpose from their saveState, or they have
-  // no implementations of either hook. Just do our thing.
-  return null;
-};
 
-/**
- * Loads the given state using either the old XML hoooks, if they should be
- * used. Returns true to indicate loading has been handled, false otherwise.
- * @param {*} callingClass The class calling this method.
- *     Used to see if `this` has overridden any relevant hooks.
- * @param {*} state The state to apply to the field.
- * @return {boolean} Whether the state was applied or not.
- */
-Field.prototype.loadLegacyState = function(callingClass, state) {
-  if (callingClass.prototype.loadState === this.loadState &&
-      callingClass.prototype.fromXml !== this.fromXml) {
-    this.fromXml(Xml.textToDom(/** @type {string} */ (state)));
-    return true;
+  /**
+   * Check whether this field's value can be changed using the editor when the
+   *     source block is editable.
+   * @return {boolean} Whether this field is enabled.
+   */
+  isEnabled() {
+    return this.enabled_;
   }
-  // Either they called this on purpose from their loadState, or they have
-  // no implementations of either hook. Just do our thing.
-  return false;
-};
-
-/**
- * Dispose of all DOM objects and events belonging to this editable field.
- * @package
- */
-Field.prototype.dispose = function() {
-  DropDownDiv.hideIfOwner(this);
-  WidgetDiv.hideIfOwner(this);
-  Tooltip.unbindMouseEvents(this.getClickTarget_());
 
-  if (this.mouseDownWrapper_) {
-    browserEvents.unbind(this.mouseDownWrapper_);
+  /**
+   * Check whether this field defines the showEditor_ function.
+   * @return {boolean} Whether this field is clickable.
+   */
+  isClickable() {
+    return this.enabled_ && !!this.sourceBlock_ &&
+        this.sourceBlock_.isEditable() &&
+        this.showEditor_ !== Field.prototype.showEditor_;
```

Comment:
I mean it is weird definitely, but idk what an alternative would be. [isClickable](https://github.com/google/blockly/blob/e21a8fcd0cf16fd26b589d6fbf42d469ca57e863/core/field.js#L705) is only used to determine if `showEditor_` should be called when the field is clicked (so we definitely want `showEditor_` to be overridden, since the base definition does nothing). I'll admit that this is also weird.

Theoretically it would be nice to refactor this, but `isClickable` is public, so I don't think we can really change its definition at this point.

Reply:
OK, that sounds reasonable enough.  I guess the "pure" design would be for this method just to return false, and subclasses that override `showEditor_` would also have to override `isClickable` to return true when applicable, but I admit that the current implementation seems more helpful.